_posts/2018-05-21-shipped-and-hosted-software.md

```diff
@@ -0,0 +1,268 @@
+  [tc-coalesce](https://github.com/mozilla/tc-coalesce) service, developed by
+  the release engineering team to support Mozilla-specific task-superseding
+  needs and hosted outside of the Taskcluster installation. On the other hand,
+  [taskcluster-stats-collector](http://github.com/taskcluster/taskcluster-stats-collector)
+  is deployed within the Firefox Taskcluster deployment, but is
+  Firefox-specific and not part of a public Taskcluster release.
+
+* While a Taskcluster release will likely encompass some pre-built worker
+  images for various cloud platforms, sophisticated worker deployment is the
+  responsibility of individual users. That may mean deploying workers to
+  hardware where necessary, perhaps with modifications to the build
+  configurations or even entirely custom-built worker implementations.  We will
+  provide cloud-provisioning tools that can be used to dynamically instantiate
+  user-specified images.
+
+#### Generated Client Libraries
+
+The second point above raises an interesting quandry: Taskcluster uses code
+generation to create its API client libraries. Historically, we have just
+pushed the "latest" client to the [package
+repository](https://yarnpkg.com/en/package/taskcluster-client) and carefully
+choreographed any incompatible changes. For users who have not customized their
+deployment, this is not too much trouble: any release of Taskcluster will have
+a client library in the package repository corresponding to it.  We don't have
+a great way to indicate which version that is, but perhaps we will invent
+something.
+
+But when Taskcluster installations are customized by adding additional
+services, progress is no longer linear: each user has a distinct "fork" of the
+Taskcluster API surface containing the locally-defined services. Development of
+Taskcluster components poses a similar challenge: if I add a new API method to
+a service, how do I call that method from another service without pushing a new
+library to the package repository?
+
+The question is further complicated by the use of compiled languages. While
+Python and JS clients can simply load a schema reference file at runtime (for
+example, a file generated at deploy time), the Go and Java clients "bake in"
+the references at compile time.
+
+Despite [much discussion](https://bugzilla.mozilla.org/show_bug.cgi?id=1428417),
+we have yet to settle on a good solution for this issue.
+
+### Everything is Public!
+
+Mozilla is Open by Design, and so is Taskcluster: with the exception of data
+that must remain private (passwords, encryption keys, and material covered by
+other companies' NDAs), everything is publicly accessible. We take advantage of
+that by reading data without any authentication. For example, the [action
+specification](https://docs.taskcluster.net/manual/using/actions/spec)
+describes downloading a decision task's `public/action.json` artifact. Nowhere
+does it mention providing any credentials to fetch the decision task, nor to
```

Comment:
This kinda makes it feel like _nothing_ is authenticated and that there isn't such a thing as credentials in taskcluster. I like the general point you're making, but maybe call out scopes/creds a bit?

Reply:
Sure, good point.